### ISSUE TITLE
Trouble uploading packages

### ISSUE THREAD

**Issue by jakirkham** [Anaconda-Platform/support#60]
Seeing an issue sporadically when trying to upload packages. The snippet below is taken from this [build log]( https://circleci.com/gh/conda-forge/statuspage-feedstock/12 ). Not quite sure why it happens or why it goes away sometimes on restart. Though it does sound like this is some sort of networking issue.

```
Using Anaconda API: https://api.anaconda.org
Traceback (most recent call last):
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 124, in <module>
    main()
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 106, in main
    on_channel = distribution_exists_on_channel(cli, meta, owner, channel)
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 69, in distribution_exists_on_channel
    binstar_cli.show_channel(owner=owner, channel=channel)['files']]
  File "/opt/conda/lib/python3.5/site-packages/binstar_client/mixins/channels.py", line 29, in show_channel
    self._check_response(res, [200])
  File "/opt/conda/lib/python3.5/site-packages/binstar_client/__init__.py", line 200, in _check_response
    raise ErrCls(msg, res.status_code)
binstar_client.errors.NotFound: ('Not Found: Nothing matches the given URI ([GET] https://api.anaconda.org/channels/conda-forge/main -> 404)', 404)
```

cc @msarahan @pelson

**Comment by jakirkham** [comment]
In some cases, am also seeing a new failure when rerunning an existing build that had already uploaded. This rerunning is being done in order to resolve some of the failed uploads. This new failure suggests the problem is not strictly limited to uploading. See the traceback below for the error and this [build]( https://ci.appveyor.com/project/conda-forge/astroid-feedstock/build/1.0.2/job/h126inb8iwx02cnx#L754 ) for more details.

```
python ci_support\upload_or_check_non_existence.py .\recipe conda-forge --channel=main
Using Anaconda API: https://api.anaconda.org
Traceback (most recent call last):
  File "ci_support\upload_or_check_non_existence.py", line 124, in <module>
    main()
  File "ci_support\upload_or_check_non_existence.py", line 106, in main
    on_channel = distribution_exists_on_channel(cli, meta, owner, channel)
  File "ci_support\upload_or_check_non_existence.py", line 69, in distribution_exists_on_channel
    binstar_cli.show_channel(owner=owner, channel=channel)['files']]
  File "C:\conda\lib\site-packages\binstar_client\mixins\channels.py", line 28, in show_channel
    res = self.session.get(url)
  File "C:\conda\lib\site-packages\requests\sessions.py", line 488, in get
    return self.request('GET', url, **kwargs)
  File "C:\conda\lib\site-packages\requests\sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\conda\lib\site-packages\requests\sessions.py", line 628, in send
    r.content
  File "C:\conda\lib\site-packages\requests\models.py", line 755, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "C:\conda\lib\site-packages\requests\models.py", line 679, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
Command exited with code 1
```

**Comment by jakirkham** [comment]
FWIW it appears this problem has been going on for the last few days (at least 3). Would really appreciate hearing that someone is looking into this even if the cause is still unknown.

**Comment by jakirkham** [comment]
Any news on this? We are still suffering from sporadic problems due to this issue.

**Comment by dsludwig** [comment]
Hi @jakirkham, I've been looking into this. It seems the API call being made in [`upload_or_check_nonexistence.py`](https://github.com/conda-forge/conda-smithy/blob/3f49bcea082d55b638f36bef2435b30ff6259c67/conda_smithy/feedstock_content/ci_support/upload_or_check_non_existence.py#L68-L69) is *really* expensive (around 20 seconds right now). As far as I can tell, the only reason it's being made is to check which labels a file has.

Right now we're working on a release that will include `labels` as an attribute returned from [the distribution api method](https://github.com/conda-forge/conda-smithy/blob/3f49bcea082d55b638f36bef2435b30ff6259c67/conda_smithy/feedstock_content/ci_support/upload_or_check_non_existence.py#L26-L27), which can be used to check the same thing, and should be much much cheaper.

I will update this issue when that has been released.

**Comment by dsludwig** [comment]
To be clear - I believe the failures are related to the expensive API operation timing out, resulting in either a broken connection or a 404, depending on where it ends up timing out.

**Comment by jakirkham** [comment]
Yeah, I'm proposing we work on eliminating that file in issue ( https://github.com/conda-forge/conda-smithy/issues/268 ) and just use `conda-build` to do the heavy lifting. Unfortunately there are two blockers. Uploading to multiple channels as described in issue ( https://github.com/conda/conda-build/issues/1230 ) and uploading to labels as described in issue ( https://github.com/conda/conda-build/issues/1229 ). Maybe this API can be tuned with `conda-build` in mind?

cc @msarahan

**Comment by jakirkham** [comment]
Thanks @dsludwig. Sorry to hear we are abusing the system. If there are other interim solutions we can try, please let us know.

**Comment by bollwyvl** [comment]
> Sorry to hear we are abusing the system

Just growing pains! Since we don't actually manage anaconda this way, conda-forge is really the first large, high-velocity user of many of these APIs! Thanks for working with us on it!

Here's what the new API will look like, now live on [http://develop.anaconda.org]( http://develop.anaconda.org/api/dist/dludwig/numpy/1.10.0/linux-64/numpy-1.10.0-py27_0.tar.bz2) (should get to beta.anaconda.org today, hopefully prod tomorrow):
```
GET /api/dist/dludwig/numpy/1.10.0/linux-64/numpy-1.10.0-py27_0.tar.bz2
```

```yaml
{
  "basename": "linux-64/numpy-1.10.0-py27_0.tar.bz2", 
  # all the other stuff...
  "labels": [
    "main"
  ], 
}
```

**Comment by jakirkham** [comment]
So should we start moving to `requests` to handle this or should we wait for an `anaconda-client` release with this functionality?

**Comment by bollwyvl** [comment]
I think it's already in there! Trying to figure out now...

**Comment by bollwyvl** [comment]
> So should we start moving to requests to handle this or should we wait for an anaconda-client release with this functionality?

This appears to work!

```python
from binstar_client import Binstar
api = Binstar(domain="http://develop.anaconda.org/api")
assert "main" in api.distribution("dludwig", "numpy", "1.10.0",
                                  "linux-64/numpy-1.10.0-py27_0.tar.bz2")["labels"]
```

**Comment by jakirkham** [comment]
Thanks guys. I'll look into how we can fix up our end at conda-forge.

**Comment by jakirkham** [comment]
Talked with @pelson earlier and we think something like PR ( https://github.com/conda-forge/conda-smithy/pull/273 ) should solve this. Please let me know your thoughts on this.

**Comment by jakirkham** [comment]
So we made a new `conda-smithy` release that breaks out that upload script. Also we have changed it to just download the index and do everything with that instead. Hopefully that fixes the issue.

Also have noticed the frequency of the problem has gone down dramatically. Guessing some things changed behind the scenes. Thanks very much for that.

In any event, closing this out. If we run into more issue, we can start a fresh issue and discuss. Thanks again for your help.
